### ISSUE TITLE
Domains mistakenly taken as TLD

### ISSUE THREAD

**Issue by OrEshed** [barseghyanartur/tld#93]
Some domains are recognized by the library as a TLD, and their subdomain to be the actual domain.

Here are some examples:

```
from tld import get_tld

get_tld('https://a.b.c.theworkpc.com', fix_protocol=True, as_object=True)
get_tld('https://www.m-o-f-365.oa.r.appspot.com', fix_protocol=True, as_object=True)
```

**Comment by Yomguithereal** [comment]
Hello @OrEshed, it seems `theworkpc.com` is a TLD as per [this](https://github.com/barseghyanartur/tld/blob/master/src/tld/res/effective_tld_names.dat.txt#L11512) line of the list. `oa.r.appspot.com` also is one as per [this](https://github.com/barseghyanartur/tld/blob/master/src/tld/res/effective_tld_names.dat.txt#L11863) line. But since they are private domains, you can still filter them out by using the `search_private` kwarg set to `True` if this is useful for your use case.

**Comment by OrEshed** [comment]
`search_private` set to `True` is the default, and it didn't work. Any other trick I can use?

**Comment by Yomguithereal** [comment]
I meant to set it to `False`, sorry :)

**Comment by OrEshed** [comment]
NP :)
I tried `False` but got this back:
`TldDomainNotFound: Domain a.b.c.theworkpc.com didn't match any existing TLD name!`
Although the domain structure is quite simple and the TLD is obviously .COM

**Comment by Yomguithereal** [comment]
@OrEshed to avoid the raised error you can set `fail_silently` to `True`.

> Although the domain structure is quite simple and the TLD is obviously .COM

No, the TLD really is `theworkpc.com` according to latest DNS data. It's possessed by [them](https://www.dynu.com/ControlPanel/AddDDNS).

**Comment by Yomguithereal** [comment]
But I agree that if we ignore private TLDs then raising an error on your case is a bit weird. @barseghyanartur do you think we should adapt the code to return `.com` in this particular case?

**Comment by OrEshed** [comment]
My problem is in obtaining the sub-domains string. I would like to get back `a.b.c`, rather than `a.b`. For my use-case, private TLDs do not count but only ICANN TLDs.

**Comment by barseghyanartur** [comment]
> But I agree that if we ignore private TLDs then raising an error on your case is a bit weird. @barseghyanartur do you think we should adapt the code to return `.com` in this particular case?

Would making a separate [BaseMozillaPublicOnlyTLDSourceParser](https://github.com/barseghyanartur/tld/blob/master/src/tld/utils.py#L172) with absolutely no information on private TLDs and then subclassing it as `MozillaPublicOnlyTLDSourceParser` solve the issue?

```python
get_tld("http://www.foreverchild", parser_class=MozillaPublicOnlyTLDSourceParser)
```

**Comment by Yomguithereal** [comment]
Could work but we could also fix the trie traversal algorithm to handle this case (while it requires backtracking I am afraid, if we don't keep specialized leaf information for the only public and only private case).

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Works in branch [feature/separate-public-only-tlds](https://github.com/barseghyanartur/tld/tree/feature/separate-public-only-tlds).

```
In [1]: from tld.utils import tld_names                                                                                                                                         

In [2]: from tld import get_tld                                                                                                                                                 

In [3]: from tld.utils import MozillaPublicOnlyTLDSourceParser                                                                                                                  

In [4]: url = 'https://a.b.c.theworkpc.com'                                                                                                                                     

In [5]: get_tld(url, parser_class=MozillaPublicOnlyTLDSourceParser)                                                                                                             
Out[5]: 'com'

In [6]: get_tld(url)                                                                                                                                                            
Out[6]: 'theworkpc.com'
```

**Comment by barseghyanartur** [comment]
@Yomguithereal:

It seems to work pretty well and performance remains almost the same. This, however, creates a tiny backwards incompatibility (in terms of behaviour), but I'll mention that in the changelog.

**Comment by Yomguithereal** [comment]
> This, however, creates a tiny backwards incompatibility (in terms of behaviour)

Indeed but since it can be seen as a bug I guess this is not a deal breaker?

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Nope. The "fix" for old behaviour, would be explicit provision of `parser_class=MozillaTLDSourceParser`, as shown in [updated tests](https://github.com/barseghyanartur/tld/pull/94/files#diff-b301641b739e312e16adc802cb9590bfedaf153fd42f78747912905fa3d48e19).

**Comment by barseghyanartur** [comment]
Fixed in 0.12.3.

**Comment by OrEshed** [comment]
Is there a need to import `parser_class=MozillaTLDSourceParser` in a special manner? I have 0.12.3 and I get a `NameError: name 'MozillaTLDSourceParser' is not defined`
